tutorials/02_functions-types.md

```diff
@@ -0,0 +1,518 @@
+
+```haskell
+a :: Integer
+b :: String
+
+a = 42
+b = a
+c = (a + 7) :: Double
+```
+
+Haskell does not have implicit conversion and if you do something like it is in the example above, you will get an error during compilation. The type system is very strict and helps programmers to find more bugs during compile time and not during runtime (where is more problematic to find it).
+
+```
+    Couldn`t match type ‘Integer’ with ‘[Char]’
+    Expected type: String
+      Actual type: Integer
+    In the expression: a
+    In an equation for ‘b’: b = a
+```
+
+### Type variable
+
+As you could have noticed, we are able to achieve polymorphism with something strange in function types, something called type variable. Type variables must start with a lowercase letter and usually are just 1 character from the beginning of the alphabet:
+
+```haskell
+identity :: a -> a
+identity x = x
+```
+
+Such type signature tells us that `identity` is a function from whatever type `a` to the same type `a`.
+
+### Type constraints
+
+In some cases (function type and others) you can see typeclass constraints like here:
+
+```haskell
+next :: Num a => a -> a
+next x = x + 1
+```
+
+It says `next` has a type `a` to `a` where `a` is "from" typeclass `Num` (typeclasses will be covered later, for now, it is just class or types). Such type signature can be even more complex.
+
+```haskell
+foo :: (Show a, Eq a, Read b) => a -> b -> a
+```
+
+### Function declaration
+
+Although type can be in most cases inferred automatically by the compiler, it is good practice to write it down at least in case of functions as part of code documentation. Functions can be complicated and by reading its type signature you know immediately what arguments it expects and what it returns.
+
+Declaration of a function is simple, just use the prefix notation as you would call the function and then describe what is it equal to.
+
+```haskell
+fibonacci   :: Word -> Integer
+fibonacci 0 = 1
+fibonacci 1 = 1
+fibonacci n = (fibonacci n-1) + (fibonacci n-2)
+```
+
+### Type declaration
+
+There are three basic ways how to introduce your own data type:
+
+1. type synonym = you just use a different name for some existing type (for example `String` is type synonym for `[Char]` = list of `Char`)
+2. new data type = declare type with type constructor (before `=`) and one or more data constructors (after `=`, separated by `|`), you may use typeclass constraints, type variables, and recursion
+3. newtype = new data type with exactly one data constructor with one parameter (new type is isomorphic with the "wrapped" type and compiler can do optimizations, can be used also in another way in more advanced code)
+
+```haskell
+type String = [Char]
+
+data Bool = True | False
+data Tree a = Leaf a | Branch (Tree a) (Tree a)
+            deriving (Show, Read)
+
+newtype Age = Age Int
+
+myTree :: Tree Int
+myTree = Branch (Leaf -7) (Branch (Leaf 7) (Leaf 10))
+
+myAge :: Age
+myAge = Age 20
+```
```

Comment:
Here, type `String` can be replaced by `[Char]` (it is a mere synonym), so typechecking is not 'perfect' in this sense. In case of `newtype`, `Age` is a different type than `Int` and compiler will check it.